[VUFIND-1677] Prepare for laminas-view v3

demiankatz opened the pull request:
Changes coming in laminas-view v3 make it impossible to extend the internals of provided view helpers, so the asset pipeline system needs to be refactored to work differently.

TODO
- [x] Create assetManager helper as wrapper around head* and foot* helpers
- [x] Revise inlineScript calls to use assetManager
  - [x] Adjust templates
  - [x] Adjust module code
  - [x] Eliminate custom InlineScript view helper
- [x] Refactor assetManager internals to support asset combination/minification when configured to do so
  - [x] Refactor CSS processing
  - [x] Refactor JS processing
- [x] Deduplicate assets when concatenating them in the pipeline
- [x] Eliminate FootScript helper
- [x] Upgrade to latest version of laminas-view
- [x] TEST!
- [x] Merge #4306 before merging this PR.
- [x] Resolve [VUFIND-1677](https://openlibraryfoundation.atlassian.net/issues/VUFIND-1677) when merging
- [x] Update changelog when merging (always use assetManager; note removal of internal helpers)

demiankatz committed:
Refactor headLink calls to assetPipeline helper.

demiankatz committed:
Move raw CSS into asset pipeline.

demiankatz committed:
Refactor headScript/footScript.

demiankatz committed:
Fix tests.

demiankatz committed:
Merge branch 'dev' into vufind-1677

demiankatz committed:
Align method parameters for consistency/completeness.

demiankatz committed:
Refactor CSS processing to reduce helper calls.

demiankatz committed:
Refactor HeadScript file collection.

demiankatz committed:
Factor CSS theme handling to AssetPipeline.

demiankatz committed:
Fix test.

demiankatz committed:
Merge branch 'dev' into vufind-1677

demiankatz committed:
Factor away HeadLink helper.

demiankatz committed:
Fix test.

demiankatz committed:
Add missing return.

demiankatz committed:
Merge branch 'dev' into vufind-1677

demiankatz committed:
Rename helper from assetPipeline to assetManager.

demiankatz committed:
Add inlineScript wrappers to AssetManager.

demiankatz committed:
Fix test.

demiankatz committed:
Revert "Fix test."

This reverts commit 92f14b5a42d2e0c653b64e5a3afd441fd0786df8.

demiankatz committed:
Update GoogleAnalytics helper to use AssetManager; improve tests.

demiankatz committed:
Fix style; port another helper.

demiankatz committed:
Finish porting helpers to use AssetManager.

demiankatz committed:
Refactor nonce handling.

demiankatz committed:
Progress on removing internal helpers.

demiankatz committed:
Only include non-empty nonce.

demiankatz committed:
Various fixes.

demiankatz committed:
Remove obsolete helper.

demiankatz committed:
Fix inline script output.

demiankatz committed:
Prevent asset pipeline duplication.

demiankatz committed:
Eliminate FootScript helper.

demiankatz committed:
Upgrade laminas-view.

demiankatz committed:
Remove redundant nonce setting.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into vufind-1677

demiankatz committed:
Factor asset pipeline logic out of asset manager.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into vufind-1677

demiankatz committed:
Merge branch 'dev' into vufind-1677

EreMaijala commented:
I've never really liked the method names used to append and prepend styles and scripts. Here's a proposal for AssetManager for maybe a bit more descriptive methods:

- appendStyleReference
- appendStyleString
- forcePrependStyleReference
- clearStyleList

- appendScriptReference
- appendScriptString
- forcePrependScriptReference
- prependScript
- clearScriptList
- outputInlineScriptReference
- outputInlineScript

Note that this proposal doesn't include an equivalent for `setStylesheet`. This would replace it: `$this->assetManager()->clearStyleList()->appendStyleReference(...)`.

EreMaijala commented on module/VuFindTheme/src/VuFindTheme/View/Helper/AssetManager.php:
```diff
@@ -0,0 +1,402 @@
+        $output = [];
+        $scriptHelper = $this->getView()->plugin('inlineScript');
+        $processedScripts = $this->pipeline->process($this->scripts[$position], 'js');
+        foreach ($processedScripts as $i => $script) {
```
$i is unused

demiankatz replied:
Fixed.

EreMaijala commented on module/VuFindTheme/src/VuFindTheme/View/Helper/AssetManager.php:
```diff
@@ -0,0 +1,402 @@
+        $processedScripts = $this->pipeline->process($this->scripts[$position], 'js');
+        foreach ($processedScripts as $i => $script) {
+            if ($script['allowArbitraryAttrs'] ?? false) {
+                $scriptHelper->setAllowArbitraryAttributes(true);
```
Should this be reset afterwards?

demiankatz replied:
Definitely! Fixed by 3e1d4eeaf900e3725dfe8a82060612b757ef09dc.

demiankatz replied:
While working on unit tests, I realized that there were some problems and inconsistencies here; addressed by 6d0231de0be1778de4ae34d926cd31ba82b67d34.

EreMaijala commented on module/VuFindTheme/src/VuFindTheme/View/Helper/AssetManager.php:
```diff
@@ -0,0 +1,402 @@
+     */
+    public function appendScript(
+        string $script,
+        string $type = 'text/javascript',
```
Since the [type attribute should not be added for JS](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type#attribute_is_not_set_default_an_empty_string_or_a_javascript_mime_type) anymore, perhaps this should not be a parameter at all in script methods. If there's any other type of script, it could be added to the attributes.

demiankatz replied:
My VM crashed while I was testing it and I'm now out of time to try again, but I'm pretty sure that 6a2df6b79e42aceeeb5d2249db552afb15a27dfc should take care of this.

demiankatz replied:
...and I finally managed to re-run tests today, and can confirm that they're still passing.

EreMaijala reviewed (changes requested):
All in all this looks pretty good to me. 

It's a bit scary as a back-compatibility break, though. I wonder if we could ease the transition by keeping HeadLink, HeadScript and InlineScript around for a little while as wrappers for AssetManager and the original helpers so that existing templates could continue to work.

demiankatz commented:
@EreMaijala, I like your proposal for method name changes. One question, though: is there a strong reason to use `Reference` in the names rather than `Link` or `URL`? (I'm fine with either option, but "Link" or "URL" is shorter if all other factors are the same). I suppose abbreviating to `Ref` is also an option, though I don't really like that idea much.

I've reached out on Slack regarding concerns about BC, just to keep the conversation here concise. I'll summarize our conclusions for the benefit of other readers once we have time to discuss.

And thanks for the other review suggestions -- I'll dig into those as soon as time permits!

EreMaijala commented:
@demiankatz `Link` instead of `Reference` would be fine for me as well. `Ref` would be even shorter though, but perhaps link is more clear. :)

demiankatz committed:
Merge branch 'dev' into vufind-1677

demiankatz committed:
Simplify  argument.

demiankatz committed:
Remove unused variable.

demiankatz committed:
Merge branch 'dev' into vufind-1677

demiankatz committed:
Reset arbitrary attributes when appropriate.

demiankatz committed:
Rename methods; add fluent interface; fix help page.

demiankatz commented:
@EreMaijala, I've adjusted the AssetManager interface in commit f7fe0e5257821d9cfca44d11576255adb105e549 to use the new method names we discussed. I also introduced a fluent interface for greater flexibility, and I fixed a problem I noticed with the help template. I think this is now ready for another review! (All tests still pass).

If this looks good to you, I can work on adding unit test coverage and preparing to backport key pieces for release 10.2. Whatever happens, I plan to leave this PR open until we can first merge the backported groundwork.

demiankatz committed:
Rename outputInlineScriptString for consistency.

demiankatz commented:
I just realized that one method was inconsistent with the rest; fixed that in b1fe578d5c111c9c90451aac10aaf2fdd52b2ec6.

demiankatz committed:
Begin work on unit test.

demiankatz committed:
Simplify and make consistent.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into vufind-1677

demiankatz committed:
Improve test; fix revealed issues.

demiankatz commented:
I made more progress on unit testing this afternoon, which revealed some bugs and interface inconsistencies that have been addressed. I ran out of time before I could finish the tests, so I'll try to do more later in the week (I covered JS logic, but CSS needs more work). Feel free to review in the meantime, but it's possible I'll fix more issues when I get to the next phase of testing.

demiankatz committed:
Finish AssetManagerTest.

EreMaijala commented on module/VuFindTheme/src/VuFindTheme/View/Helper/AssetManager.php:
```diff
@@ -0,0 +1,426 @@
+    protected function outputScriptAssets($position): string
+    {
+        $output = [];
+        $scriptHelper = $this->getView()->plugin('inlineScript');
```
```suggestion
```

demiankatz committed:
Update module/VuFindTheme/src/VuFindTheme/View/Helper/AssetManager.php

Co-authored-by: Ere Maijala <ere.maijala@helsinki.fi>

demiankatz commented:
Good catch; thanks, @EreMaijala!

demiankatz committed:
Merge branch 'dev' into vufind-1677

demiankatz committed:
Add exclude_from_pipeline option.

demiankatz committed:
Add missing clear to help page.

demiankatz committed:
Eliminate allowArbitraryAttributes argument.

demiankatz committed:
Merge branch 'dev' into vufind-1677

demiankatz committed:
Merge branch 'dev' into vufind-1677

demiankatz committed:
Fix minor configuration processing bug; add test.

demiankatz committed:
More test progress; minor refactoring.

demiankatz committed:
Add CSS tests.

demiankatz committed:
Test group processing.

demiankatz committed:
Merge branch 'dev' into vufind-1677
